From a886524590b5e980617c67be49deb941f9ece16c Mon Sep 17 00:00:00 2001
From: Claudiu Manoil <claudiu.manoil@nxp.com>
Date: Mon, 9 Oct 2017 15:23:42 +0300
Subject: [PATCH 064/706] enetc: Enable VLAN promisc mode for SI0 by default

This should have been a HW default, to allow VLAN tagged
traffic to enter the system, whithout having to configure
VLAN h/w acceleration (i.e. h/w VLAN filters) for that.

Fixes unwanted behavior below (m0165 sim model):
Port[0] MAC[0] ++++ SI0 : Performing VLAN filtering ...
Port[0] MAC[0] SI0 : VLAN promiscuous mode is not set
Port[0] MAC[0] SI0 : Frame is tagged (VLAN1 present)
Port[0] MAC[0] SI0 : Reading outer VLAN param : pcp = 0x0, dei = 0, vid
= 0x2, tpid = 0x0
Port[0] MAC[0] SI0 : isolation vid = 0x0, isolation tpid = 0x0
Port[0] MAC[0] SI0 : Received vid/tpid don't match with isolation
vid/tpid
Port[0] MAC[0] SI0 : Tagged traffic will go through VLAN filter table
Port[0] MAC[0] SI0 : Searching in VLAN filtering table for vid = 0x2
Port[0] MAC[0] SI0 : Completed search in VLAN filtering table for vid =
0x2, vlan_si_select[0] = 0
Port[0] MAC[0] ++++ SI0 VLAN Filtering result : vlan_si_select[0] = 0
[...]
Port[0] MAC[0] SI0 not selected, discarding rx packet for this SI!

Signed-off-by: Claudiu Manoil <claudiu.manoil@nxp.com>
(cherry picked from commit 06cfdba45850c445b94eddaf339f8b9972496419)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c    | 2 ++
 drivers/net/ethernet/freescale/enetc/enetc_hw.h | 1 +
 2 files changed, 3 insertions(+)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index f672006c1963..bd62f7b79b66 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -1262,6 +1262,8 @@ static void enetc_configure_port(struct enetc_ndev_priv *priv)
 	enetc_port_wr(hw, ENETC_PMR, ENETC_PMR_EN);
 
 	enetc_port_setup_primary_mac_address(priv);
+	/* reset promiscuity to default values, except VLAN promisc for SI0 */
+	enetc_port_wr(hw, ENETC_PSIPMR, ENETC_PSIPMR_SET_VP(0));
 }
 
 static void enetc_configure_si(struct enetc_ndev_priv *priv)
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_hw.h b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
index a14d9ffddda7..2a4aa0d2b4af 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_hw.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
@@ -76,6 +76,7 @@ enum enetc_bdr_type {TX, RX};
 #define ENETC_PSIPMR	0x00018
 #define ENETC_PSIPMR_SET_UP(n)	(0x1 << (n)) /* n = SI index */
 #define ENETC_PSIPMR_SET_MP(n)	(0x1 << ((n) + 8))
+#define ENETC_PSIPMR_SET_VP(n)	(0x1 << ((n) + 16))
 #define ENETC_PSIPMAR0(n)	(0x00100 + (n) * 0x20) /* n = SI index */
 #define ENETC_PSIPMAR1(n)	(0x00104 + (n) * 0x20)
 #define ENETC_PSIIVLANR(n)	(0x00210 + (n) * 4) /* n = SI index */
-- 
2.17.1

